test/runtime/Policies.go

```diff
@@ -486,6 +486,33 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
 		connectivityTest(allRequests, helpers.App2, helpers.Httpd1, true)
 	})
 
+	It("Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated", func() {
+		httpd1DockerNetworking, err := vm.ContainerInspectNet(helpers.Httpd1)
+
+		By("Importing policy and waiting for revision to increase for endpoints")
+		_, err = vm.PolicyImportAndWait(vm.GetFullPath(policiesL7JSON), helpers.HelperTimeout)
+		Expect(err).ToNot(HaveOccurred(), "unable to import policy after timeout")
+
+		By("Trying to access %s:80/public from %s (should be allowed by policy)", helpers.Httpd1, helpers.App1)
+		res := vm.ContainerExec(helpers.App1, helpers.CurlFail("http://%s:80/public", fmt.Sprintf(httpd1DockerNetworking[helpers.IPv4])))
+		res.ExpectSuccess("unable to access %s:80/public from %s (should have worked)", helpers.Httpd1, helpers.App1)
+
+		By("Changing daemon configuration")
+		res = vm.ExecCilium("config Debug=true")
+		res.ExpectSuccess("unable to change daemon configuration")
+
+		res = vm.ExecCilium("config Debug=false")
+		res.ExpectSuccess("unable to change daemon configuration")
+
+		By("Waiting for endpoints to be ready after daemon configuration change")
```

Comment:
How does it test that an endpoint generation has actually happened?

Reply:
So I went back and manually verified that this test failed if I checked out master at a specific commit from before this fix was made. I did modify the test to not wait for policy revision increases for the endpoints because prior to my change, policy revision wasn't bumped upon daemon configuration updates:

``` 
$ git checkout master && git checkout 7d8f833848b1449147a54063118a3c9ee17c0ba2
$ git cherry-pick 0eb3a1fa87fbffdcb641b9a503e6bc862f54b684
[detached HEAD f4fb96a16] test/runtime: add endpoint regeneration test
 Date: Tue May 29 14:30:15 2018 -0700
 1 file changed, 57 insertions(+)
$ cd test/ && vagrant up runtime
$ ginkgo --focus="RuntimeValidatedPolicies" --focus="Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated" -v -- -cilium.provision=false -cilium.holdEnvironment=true
level=info msg="environment variable \"K8S_VERSION\" was not set; setting to default value \"1.10\""
Running Suite: runtime
======================
Random Seed: 1527808210
Will run 1 of 288 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
RuntimeValidatedPolicies 
  Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated
  /Users/ianvernon/go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:345
STEP: Setting PolicyEnforcement=default
STEP: Importing policy and waiting for revision to increase for endpoints
STEP: Setting up policy: /home/vagrant/go/src/github.com/cilium/cilium/test/runtime/manifests/Policies-l7-simple.json
STEP: Trying to access httpd1:80/public from app1 (should be allowed by policy)
STEP: Getting configuration for daemon
STEP: Changing daemon configuration from Debug=Disabled to Debug=true to induce policy recalculation for endpoints
STEP: Waiting for policy revision to increase after daemon configuration change
STEP: Changing daemon configuration from Debug=true to Debug=Disabled to induce policy recalculation for endpoints
STEP: Trying to access httpd1:80/public from app1 (should be allowed by policy)

---
RuntimeValidatedPolicies Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated
at /Users/ianvernon/go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:345

unable to access httpd1:80/public from app1 (should have worked)
Expected command: docker exec -i  app1 curl -s --fail --connect-timeout 5 --max-time 5 http://10.15.148.173:80/public 
To succeed, but it failed:
Exitcode: 28 
Stdout:
 	 
Stderr:
 	 


Pausing test for debug, use vagrant to access test setup.
Run "kill -SIGCONT 81573" to continue.
^Z
[1]+  Stopped                 ginkgo --focus="RuntimeValidatedPolicies" --focus="Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated" -v -- -cilium.provision=false -cilium.holdEnvironment=true
Ians-MBP:test ianvernon$ fg
ginkgo --focus="RuntimeValidatedPolicies" --focus="Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated" -v -- -cilium.provision=false -cilium.holdEnvironment=true
Skipped gathering logs (-cilium.holdEnvironment=true)


• Failure [171.087 seconds]
RuntimeValidatedPolicies
/Users/ianvernon/go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:313
  Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated [It]
  /Users/ianvernon/go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:345

  unable to access httpd1:80/public from app1 (should have worked)
  Expected command: docker exec -i  app1 curl -s --fail --connect-timeout 5 --max-time 5 http://10.15.148.173:80/public 
  To succeed, but it failed:
  Exitcode: 28 
  Stdout:
   	 
  Stderr:
   	 
  

  /Users/ianvernon/go/src/github.com/cilium/cilium/test/runtime/Policies.go:542
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS

Summarizing 1 Failure:

[Fail] RuntimeValidatedPolicies [It] Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated 
/Users/ianvernon/go/src/github.com/cilium/cilium/test/runtime/Policies.go:542

Ran 1 of 288 Specs in 171.090 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 287 Skipped --- FAIL: TestTest (171.11s)
FAIL

Ginkgo ran 1 suite in 3m24.310484748s
Test Suite Failed
```

I then cherry-picked the changes related to #4215 : 

```
$ git cherry-pick 8d72fd92ae042700d6ce4e05a4988e32011fe9c7
[detached HEAD a461fa06c] pkg/endpoint: include node headerfile hash
 Date: Wed May 23 19:11:56 2018 -0700
 2 files changed, 66 insertions(+), 29 deletions(-)
$ git cherry-pick 3b9b744a5fc0e4f8b7d95d28017665c81b4d6c3f
[detached HEAD ac235b39e] pkg/endpointmanager: always regenerate if policy forcibly computed
 Date: Wed May 23 18:54:12 2018 -0700
 2 files changed, 4 insertions(+), 4 deletions(-)
$ git cherry-pick 659d5acfa8c54f1257d3d57ab8d5732109ee1cca
[detached HEAD b7b0756d0] daemon: trigger policy updates upon daemon configuration update
 Date: Wed May 23 19:35:55 2018 -0700
 1 file changed, 1 insertion(+), 2 deletions(-)
```

Then SSH'd into the VM, recompiled Cilium, and ran the test again:

```
$ vagrant ssh runtime
$ make && sudo make install && sudo service cilium restart
$ exit
$ $ ginkgo --focus="RuntimeValidatedPolicies" --focus="Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated" -v -- -cilium.provision=false -cilium.holdEnvironment=true
level=info msg="environment variable \"K8S_VERSION\" was not set; setting to default value \"1.10\""
Running Suite: runtime
======================
Random Seed: 1527809151
Will run 1 of 288 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
RuntimeValidatedPolicies 
  Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated
  /Users/ianvernon/go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:345
STEP: Setting PolicyEnforcement=default
STEP: Importing policy and waiting for revision to increase for endpoints
STEP: Setting up policy: /home/vagrant/go/src/github.com/cilium/cilium/test/runtime/manifests/Policies-l7-simple.json
STEP: Trying to access httpd1:80/public from app1 (should be allowed by policy)
STEP: Getting configuration for daemon
STEP: Changing daemon configuration from Debug=Disabled to Debug=true to induce policy recalculation for endpoints
STEP: Waiting for policy revision to increase after daemon configuration change
STEP: Changing daemon configuration from Debug=true to Debug=Disabled to induce policy recalculation for endpoints
STEP: Trying to access httpd1:80/public from app1 (should be allowed by policy)

• [SLOW TEST:50.520 seconds]
RuntimeValidatedPolicies
/Users/ianvernon/go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:313
  Tests Endpoint Regeneration Occurs When Daemon Configuration Is Updated
  /Users/ianvernon/go/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:345
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSS
Ran 1 of 288 Specs in 50.523 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 287 Skipped PASS

Ginkgo ran 1 suite in 59.268953143s
Test Suite Passed
```

Thus, if endpoint regeneration had not occurred, the test would have failed. However,  I've changed the name of the `It` to reflect that this tests connectivity as opposed to regeneration.